data/tutorials/language/1ms_01_functors.md

```diff
@@ -8,183 +8,392 @@ category: "Module System"
+
+```shell
+$ dune exec funkt < dune
+executable
+libraries
+name
+public_name
+str
+funkt
+```
+
+There are no duplicates in a `Set`. Therefore, the string `"funkt"` is only displayed once, although it appears twice in the `dune` file.
+
+## Extending a Module with a Standard Library Functor
+
+Using the `include` statement, here is an alternate way to expose the module created by `Set.Make(String)`:
+
+**`funkt.ml`**
+```ocaml
+module String = struct
+  include String
+  module Set = Set.Make(String)
+end
+
+let _ =
+  stdin
+  |> In_channel.input_lines
+  |> List.concat_map Str.(split (regexp "[ \t.,;:()]+"))
+  |> String.Set.of_list
+  |> String.Set.iter print_endline
+```
+
+This allows the user to seemingly extend the module `String` with a submodule `Set`. Check the behaviour using `dune exec funkt < dune`.
+
+## Functors are Parametrised Modules
+
+### Functors from the Standard Library
+
+Some ”modules” provide operations over an abstract type and need to be supplied with a parameter module used in their implementation. These “modules” are parametrised, in other words, functors. That's the case for the sets, maps, and hash tables provided by the standard library. It works in a contract way:
+* If you provide a module that implements what is expected (the parameter interface)
+* The functor returns a module that implements what is promised (the result interface)
+
+Here is the signature of the module that the functors `Set.Make` and `Map.Make` expect:
+```ocaml
+module type OrderedType = sig
+  type t
+  val compare : t -> t -> int
+end
+```
+
+Here is the signature of the module that the functor `Hashtbl.Make` expects:
+```ocaml
+module type HashedType = sig
+  type t
+  val equal : t -> t -> bool
+  val hash : t -> int
+end
+```
+
+**Note**: `Ordered.t` is a type of set elements or map keys, `Set.S.t` is a type of set, and `Map.S.t` is a type of mapping. `HashedType.t` is a type of hash table keys, and `Hashtbl.S.t` is a type of hash table.
+
+The functors  `Set.Make`, `Map.Make`, and `Hashtbl.Make` return modules satisfying the interfaces `Set.S`, `Map.S`, and `Hashtbl.S` (respectively), which all contain an abstract type `t` and associated functions. Refer to the documentation for the details about what they provide:
+* [`Set.S`](/api/Set.S.html)
+* [`Map.S`](/api/Map.S.html)
+* [`Hashtbl.S`](/api/Hashtbl.S.html)
+
+### Writing Your Own Functors
+
+There are many kinds of [heap](https://en.wikipedia.org/wiki/Heap_(data_structure)) data structures. Example include binary heaps, leftist heaps, binomial heaps, or Fibonacci heaps.
+
+The kind of data structures and algorithms used to implement a heap is not discussed in this document.
+
+The common prerequisite to implementing any kind of heap is the availability of a means to compare the elements they contain. That's the same signature as the parameter of `Set.Make` and `Map.Make`:
```

Comment:
```suggestion
The common prerequisite to implement any heap is the available means to compare the elements contained in that heap. That's the same signature as the parameter of `Set.Make` and `Map.Make`:
```

If this isn't correct, let's work on clarifying and tightening this sentence together.

Reply:
I have shortened it further

Reply:
Nicely done!